src/hotspot/share/runtime/thread.cpp

```diff
@@ -938,14 +941,19 @@ void Thread::print_on_error(outputStream* st, char* buf, int buflen) const {
     st->print(" \"%s\"", name());
   }
 
-  st->print(" [stack: " PTR_FORMAT "," PTR_FORMAT "]",
-            p2i(stack_end()), p2i(stack_base()));
-
-  if (osthread()) {
-    st->print(" [id=%d]", osthread()->thread_id());
+  OSThread* os_thr = osthread();
+  if (os_thr != NULL) {
+    if (os_thr->get_state() != ZOMBIE) {
+      st->print(" [stack: " PTR_FORMAT "," PTR_FORMAT "]",
+                p2i(stack_end()), p2i(stack_base()));
+      st->print(" [id=%d]", osthread()->thread_id());
+      ThreadsSMRSupport::print_info_on(this, st);
+    } else {
+      st->print(" Terminated");
+    }
+  } else {
+    st->print(" Aborted");
```

Comment:
Not sure this is reachable and if it is then I'm not sure what state the thread is actually in. If a Thread never gets an osThread() it isn't started so shouldn't be locatable by any means.

Reply:
so, you prefer "ShouldNotReachHere()" ?

Reply:
There's no point putting a ShouldNotReachHere() in error handling code as we will just trip a secondary error.
If we want to print something the perhaps "unknown state (no osThread)" ?
Also I only wanted the ThreadSMRSupport::print_info_on to be excluded for Zombies, but you've excluded it for the no-osThread case as well. I think based on what Dan said we can just put that back and call it unconditionally. Thanks.

Reply:
+1 for "unknown state". "Aborted" is misleading since we do not know. Could have crashed right at thread creation (or is that what "Aborted" means?)